PEP: Uniform URLs in core metadata (#1)
woodruffw-forks/peps: peps/pep-9999.rst

choosing to provide only a ``project.urls`` table rather than a
``project.home-page`` field. From :pep:`621`'s rejected ideas:

.. pull-quote::

    While the core metadata supports it, having a single field for a project's
    URL while also supporting a full table seemed redundant and confusing.

This PEP exists to formalize the informal conventions that have arisen, as well
as explicitly document ``Home-page`` and ``Download-URL`` as deprecated in
favor of equivalent ``Project-URL`` representations.

Deprecating ``Home-page`` and ``Download-URL``
==============================================

This PEP proposes that ``Home-page`` and ``Download-URL`` be considered
**deprecated**. This deprecation has implications for both package metadata
producers (e.g. build backends and packaging tools) and package indices
(e.g. PyPI).

.. _metadata-producers:

Metadata producers
------------------

This PEP stipulates the following for metadata producers:

* When generating metadata 1.2 or layer, producers **SHOULD** emit only
  ``Project-URL``, and **SHOULD NOT** emit ``Home-page`` or ``Download-URL``
  fields.
* When generating ``Project-URL`` equivalents for ``Home-page`` and

>   ``Download-URL``, producers **SHALL** use the

Reviewer: `SHALL` isn't a common term for requirements in my experience - maybe it would be better as `MUST` (if that's the intended meaning)?

Author: Thanks! `SHALL` is listed in RFC 2119 as an equivalent of `MUST`, but agreed that it's not common. Changing!

Author: (I also realized this particular `SHALL` should be a `SHOULD`.)

conventions have arisen for representing the values of
``Home-page`` and ``Download-URL`` within ``Project-URL`` instead.

These conventions have seen significant adoption, with :pep:`621` explicitly
choosing to provide only a ``project.urls`` table rather than a
``project.home-page`` field. From :pep:`621`'s rejected ideas:

.. pull-quote::

    While the core metadata supports it, having a single field for a project's
    URL while also supporting a full table seemed redundant and confusing.

This PEP exists to formalize the informal conventions that have arisen, as well
as explicitly document ``Home-page`` and ``Download-URL`` as deprecated in
favor of equivalent ``Project-URL`` representations.

Deprecating ``Home-page`` and ``Download-URL``
==============================================

This PEP proposes that ``Home-page`` and ``Download-URL`` be considered
**deprecated**. This deprecation has implications for both package metadata
producers (e.g. build backends and packaging tools) and package indices
(e.g. PyPI).

.. _metadata-producers:

Metadata producers
------------------

This PEP stipulates the following for metadata producers:

> * When generating metadata 1.2 or layer, producers **SHOULD** emit only

Reviewer 2: ```suggestion
* When generating metadata 1.2 or later, producers **SHOULD** emit only
```

* When interpreting a distribution's metadata of version 1.2 or later
  (e.g. for rendering on a web page), the index **SHOULD** prefer
  ``Project-URL`` fields as a source of URLs over ``Home-page`` and
  ``Download-URL``, even if the latter are explicitly provided.

* If a distributions's metadata contains only the ``Home-page`` and
  ``Download-URL`` fields, the index **MAY** choose to ignore those fields
  and behave as though no URLs were provided in the metadata. In this case,
  the index **SHOULD** present an appropriate warning or notification to
  the uploading party.

  * The mechanism for presenting this warning or notification is not
    specified, since it will vary by index. By way of example, an index may
    choose to present a warning in the HTTP response to an upload request, or
    send an email or other notification to the maintainer(s) of the project.

* If a distribution's metadata contains both sets of fields, the index **MAY**
  choose to reject the distribution outright. However, this is
  **NOT RECOMMENDED** until a future unspecified major metadata version
  formally removes support for ``Home-page`` and ``Download-URL``.

> These stipulations do not change the optionality of URL processing by indices.

Reviewer 2: Ideally, package repositories would refrain from reinterpreting the URLs in previously uploaded packages in ways that lose information:

```suggestion

* Any changes to the interpretation of metadata of version 1.2 or later that
  result in previously recognized URLs no longer being recognized
  **SHOULD NOT** be retroactively applied to previously uploaded packages.

These stipulations do not change the optionality of URL processing by indices.
```

In other words, an index that does not process URLs within uploaded
distributions may continue to ignore all URL fields entirely.

.. _conventions-for-labels:

Conventions for ``Project-URL`` labels
======================================

The deprecations proposed above require a formalization of the currently
informal relationship between ``Home-page``, ``Download-URL``, and their
``Project-URL`` equivalents.

This formalization has two parts:

1. A set of rules for canonicalizing ``Project-URL`` labels;
2. A set of "well-known" canonical label values that indices may specialize
   URL presentation for.

Label canonicalization
----------------------

The core metadata specification stipulates that ``Project-URL`` labels are
free text, limited to 32 characters.

This PEP proposes adding the concept of a "canonicalized" label to the core
metadata specification. Label canonicalization is defined via the following
Python function:

.. code-block:: python

>     def canonicalize_label(label: str) -> str:
>         return label.lower().replace("-", "").replace("_", "").replace(" ", "")

Reviewer 2: Seems simpler to me to just remove all ASCII punctuation and whitespace rather than doing it selectively:

```suggestion
    import string
    def canonicalize_label(label: str) -> str:
        chars_to_remove = string.punctuation + string.whitespace
        removal_map = str.maketrans("", "", chars_to_remove)
        return label.translate(removal_map).lower()
```

(e.g. `What's New?` would leave the `'` and `?` behind under the initial draft rules)

Author: Thank you, that's way simpler!

> In plain language: a label is *canonicalized* by deleting all spans of dashes,
> underscores, and spaces.

Reviewer 2: ```suggestion
In plain language: a label is *canonicalized* by deleting all ASCII punctuation and
whitespace, and then converting the result to lowercase.
```

The following table shows examples of labels before (raw) and after
canonicalization:

.. csv-table::
    :header: "Raw", "Canonicalized"

    "``Homepage``", "``homepage``"
    "``Home-page``", "``homepage``"
    "``Home page``", "``homepage``"
    "``Change_Log``", "``changelog``"

Metadata producers **SHOULD** emit the canonicalized form of a user
specified label, but **MAY** choose to emit the un-canonicalized form so
long as it adheres to the existing 32 character constraint.

> 

Reviewer 2: ```suggestion

Package indices *SHOULD NOT** use the canonicalized labels directly as UI
elements (instead replacing them with appropriately capitalized text labels).

```

Well-known labels
-----------------

In addition to the canonicalization rules above, this PEP proposes a
fixed (but extensible) set of "well-known" ``Project-URL`` labels,
as well as equivalent aliases.

The following table lists these labels, in canonical form:

.. csv-table::
    :header: "Label", "Description", "Aliases"
    :widths: 20, 60, 20

    "``homepage``", "The project's homepage", "*(none)*"
    "``download``", "A download URL for the current distribution, equivalent to ``Download-URL``", "*(none)*"
    "``changelog``", "The project's changelog", "``changes``, ``releasenotes``, ``whatsnew``, ``history``"
    "``documentation``", "The project's online documentation", "``docs``"
    "``bug``", "The project's bug tracker", "``issue``, ``tracker``, ``report``"
    "``funding``", "Sponsoring information", "``sponsor``, ``donate``, ``donation``"

Packagers and metadata producers **MAY** choose to use these well-known
labels to communicate specific URL intents to package indices and downstreams.

Similarly, indices **MAY** choose to specialize their rendering or presentation
of URLs with these labels, e.g. by presenting an appropriate icon or tooltip
for each label.

> 

Reviewer 2: A possible way to capture the fact PyPI's labelling goes a bit further than this, without actually spelling out the exact details of those extensions at the standardisation level.

```suggestion

Indices **MAY** also specialize the rendering or presentation of additional labels or URLs,
including (but not limited to), labels that start with a well-known label, and URLs that refer
to a known service provider domain (e.g. for documentation hosting or issue tracking).

```

This PEP recognizes that the list of well-known labels is unlikely to remain
static, and that subsequent additions to it should not require the overhead

> associated with a formal PEP process or new metadata version. Consequently,
> this PEP proposes that the list above become a "living" list within the
> living core metadata standard, which PyPA members may amend via ordinary
> contributing processes.

Reviewer 2: Rather than doing that, perhaps we should delegate the role of documenting the living conventions specifically to the Python Package Index documentation? While we normally avoid letting implementations control interoperability, PyPI does hold a genuinely privileged position in this case (and we'd be delegating this to *PyPI* as service, not technically to `warehouse` as an implementation).

```suggestion
associated with a formal PEP process or new metadata version. As the primary
expected use case for this information is to control the way project URLs are
displayed on the Python Package Index, this PEP proposes that the list above
become a "living" list within PyPI's documentation (at time of writing, the
documentation for influencing PyPI's URL display can be found
`here <https://docs.pypi.org/project_metadata/#icons>`__).
```

Backwards Compatibility
=======================

Limited Impact
--------------

This PEP is expected to have little to no impact on existing packaging tooling
or package indices:

* Packaging tooling: no changes to the correctness or well-formedness
  of the core metadata. This PEP proposes deprecations as well as behavioral
  refinements, but all currently (and historically) produced metadata will
  continue to be valid per the rules of its respective version.

> * Package indices: indices will continue to except well-formed core metadata,

Reviewer 2: ```suggestion
* Package indices: indices will continue to expect well-formed core metadata,
```

  with no behavioral changes. Indices **MAY** choose to emit warnings or
  notifications on the presence of now-deprecated fields,
  :ref:`per above <package-indices>`.

.. _future-considerations:

Future Considerations
=====================

This PEP does not stipulate or require any future metadata changes.

However, per :ref:`metadata-producers` and :ref:`conventions-for-labels`,
we identify the following potential future goals for a new major release of
the core metadata standards:

* Outright removal of support for ``Home-page`` and ``Download-URL`` in the
  next major core metadata version. If removed, package indices and consumers
  **MUST** reject metadata containing these fields when said metadata is of
  the new major version.
* Enforcement of label canonicalization. If enforced, package producers
  **MUST** emit only canonicalized ``Project-URL`` labels when generating

>   distribution metadata, and package indices and consumers **MUST** reject
>   distributions containing non-canonicalized labels.

Reviewer 2: ```suggestion
  distribution metadata, and package indices and consumers **MUST** reject
  distributions containing non-canonicalized labels. Note: requiring
  canonicalization merely restricts labels to lowercase text, and excludes
  whitespace and punctuation. It does NOT restrict project URLs solely to
  the use of "well-known" labels. 
```

> These potential changes would be backwards incompatible, hence their
> inclusion only in this section.

Reviewer 2: This is already said above, but it seems worth repeating here:

```suggestion
These potential changes would be backwards incompatible, hence their
inclusion only in this section. Acceptance of this PEP does NOT commit
any future metadata revision to actually making these changes.
```
